### ISSUE TITLE
False positive on calls to bound methods of classes when using a typevar bound to a union of those classes

### ISSUE THREAD

**Issue by Wizzerinus** [astral-sh/ty#2585]
### Summary

Consider the following example:

```py
from typing import TypeVar


class A:
    def callMe(self):
        return 1


class B:
    def callMe(self):
        return None


T = TypeVar("T", bound=A | B)


def perform2(x: T):
    x.callMe()
```

If we have an object of type `A | B`, we can safely call `callMe()` on it. ty agrees.

However, if we have an object of type `T` where `T` is bound to `A | B`, ty declares that we can't all `callMe`:

https://play.ty.dev/3ef85a63-f4a0-4ba4-a9f4-18e9b7ebcf2b

Seems related to #1503 , but that one is about constrained typevars, not bound ones.

### Version

0.0.13

**Comment by tibdex** [comment]
Another example:

```py
from dataclasses import dataclass
from typing import Generic, Literal, TypeAlias, TypeVar

from typing_extensions import reveal_type

MembershipOperator: TypeAlias = Literal["IN", "NOT_IN"]
```

Expected behavior:

```py
@dataclass
class Condition:
    operator: MembershipOperator

    def __repr__(self) -> str:
        reveal_type(self.operator)  # Literal["IN", "NOT_IN"]

        match self.operator:  # ty sees exhaustiveness
            case "IN":
                return "in_result"
            case "NOT_IN":
                return "not_in_result"
```

Unexpected behavior:

```py
OperatorT = TypeVar("OperatorT", bound=MembershipOperator)

@dataclass
class GenericCondition(Generic[OperatorT]):
    operator: OperatorT

    def __repr__(self) -> str:
        reveal_type(self.operator) # Unknown

        match self.operator: # ty does not see exhaustiveness
            case "IN":
                return "in_result"
            case "NOT_IN":
                return "not_in_result"
```

`ty` raises:

> Function can implicitly return `None`, which is not assignable to return type `str`

because it considered the `match` statement to not be exhaustive which is not what I expected.

---

python: 3.10.16
ty: 0.0.16

**Comment by carljm** [comment]
@tibdex That example is not this issue (it doesn't involve calling methods), it's #2726

**Comment by tibdex** [comment]
Sorry Carl for being off-topic, I looked at the issue summary rather than its title and it did seem closely related (involving attribute access rather than method call). I didn't find a more suitable issue in [the existing issues](https://github.com/astral-sh/ty/issues?q=is%3Aissue%20state%3Aopen%20typevar%20union&page=1). Should I create a new issue or can I subscribe to this one and hope that its fix will also fix my issue?
Thank you.

**Comment by carljm** [comment]
@tibdex I think your issue is #2726 

and no worries, the off-topic hide is not trying to say you did something wrong, it's just trying to avoid distraction/confusion for someone trying to fix this issue

**Comment by carljm** [comment]
Another manifestation of this (from #3621):

```py
def convert_data[T: float](data: T):
    data / 1  # ty errors here (but not if we just `JustFloat` instead of `float`, or if we remove the typevar)
```
